src/WebJobs.Script.WebHost/FunctionRequestInvoker.cs

```diff
@@ -0,0 +1,122 @@
+            }
+        }
+
+        public static async Task<HttpResponseMessage> ProcessRequestAsync(HttpRequestMessage request, FunctionDescriptor function, CancellationToken cancellationToken, WebScriptHostManager scriptHostManager, WebHookReceiverManager webHookReceiverManager)
+        {
+            var httpTrigger = function.GetTriggerAttributeOrNull<HttpTriggerAttribute>();
+            bool isWebHook = !string.IsNullOrEmpty(httpTrigger.WebHookType);
+            var authorizationLevel = request.GetAuthorizationLevel();
+            HttpResponseMessage response = null;
+
+            if (isWebHook)
+            {
+                if (request.HasAuthorizationLevel(AuthorizationLevel.Admin))
+                {
+                    // Admin level requests bypass the WebHook auth pipeline
+                    response = await scriptHostManager.HandleRequestAsync(function, request, cancellationToken);
+                }
+                else
+                {
+                    // This is a WebHook request so define a delegate for the user function.
+                    // The WebHook Receiver pipeline will first validate the request fully
+                    // then invoke this callback.
+                    Func<HttpRequestMessage, Task<HttpResponseMessage>> invokeFunction = async (req) =>
+                    {
+                        // Reset the content stream before passing the request down to the function
+                        Stream stream = await req.Content.ReadAsStreamAsync();
+                        stream.Seek(0, SeekOrigin.Begin);
+
+                        return await scriptHostManager.HandleRequestAsync(function, req, cancellationToken);
+                    };
+                    response = await webHookReceiverManager.HandleRequestAsync(function, request, invokeFunction);
+                }
+            }
+            else
+            {
+                // Authorize
+                if (!request.HasAuthorizationLevel(httpTrigger.AuthLevel))
+                {
+                    return new HttpResponseMessage(HttpStatusCode.Unauthorized);
+                }
+
+                // Not a WebHook request so dispatch directly
+                response = await scriptHostManager.HandleRequestAsync(function, request, cancellationToken);
+            }
+
+            return response;
+        }
+
+        public static async Task<HttpResponseMessage> PreValidateRequestAsync(HttpRequestMessage request, FunctionDescriptor function, IDependencyResolver resolver)
+        {
+            if (function == null)
+            {
+                if (request.RequestUri.AbsolutePath == "/")
+                {
+                    // if the request is to the root and we can't find any matching FunctionDescriptors which might have been setup by proxies
+                    // then homepage logic will be applied.
+                    return (IsHomepageDisabled || request.IsAntaresInternalRequest())
```

Comment:
The naming of this method isn't correct, since here we're doing more than validation - we're returning the home page

Reply:
Good point, I am going to rename it to PreProcessRequestAsync. Proxy is just returning what its route is configured to return. imagine a scenario where the route is to a local funaction and in request overrides it gets changed to rhe root of the function app. we just need to be consistent and make sure the same code path runs for both.